File: codeabb/C/p104_triangle_area.c
// codeabbey, problem 104
// solution by whoisrgj

#include <stdio.h>
#include <math.h>

int main(int argc, char *argv[]) {
    int N, i;
    double a, b, c, s, area;
    
    struct point {
        int x;
        int y;
    };
    struct point v[3];
    
    fscanf(stdin, "%d", &N);
    for(i = 0; i < N; i++) {
        fscanf(stdin, "%d %d %d %d %d %d", &v[0].x, &v[0].y, &v[1].x, &v[1].y, &v[2].x, &v[2].y);
        a = sqrt(pow(v[1].x-v[0].x, 2)+pow(v[1].y-v[0].y, 2));
        b = sqrt(pow(v[2].x-v[1].x, 2)+pow(v[2].y-v[1].y, 2));
        c = sqrt(pow(v[0].x-v[2].x, 2)+pow(v[0].y-v[2].y, 2));
        s = (a+b+c)/2;
        area = sqrt(s*(s-a)*(s-b)*(s-c));
        fprintf(stdout, "%.8g ", area);
    }
    return 0;
}

File: codeabb/Python/p104_triangle_area.py
"""
CodeAbbey, Problem 104
Coded by whoisrgj
"""
from collections import namedtuple
import math

# using namedtuple for practice
Point = namedtuple('Point', 'x y')


def triangle_area(p1: Point, p2: Point, p3: Point):
    a = math.sqrt(math.pow(p2.x - p1.x, 2) + math.pow(p2.y - p1.y, 2))
    b = math.sqrt(math.pow(p3.x - p2.x, 2) + math.pow(p3.y - p2.y, 2))
    c = math.sqrt(math.pow(p1.x - p3.x, 2) + math.pow(p1.y - p3.y, 2))
    s = (a + b + c) / 2
    return "{:.8g}".format(math.sqrt(s * (s - a) * (s - b) * (s - c)))


# formula from author's notes, abs value of the vector product divided by 2
def triangle_area2(p1: Point, p2: Point, p3: Point):
    p = p1.x * (p2.y - p3.y) + p2.x * (p3.y - p1.y) + p3.x * (p1.y - p2.y)
    return "{:.8g}".format(abs(p) / 2)


def main():
    n = int(input())
    results = []
    for _ in range(n):
        x1, y1, x2, y2, x3, y3 = map(int, input().split())
        p1, p2, p3 = Point(x1, y1), Point(x2, y2), Point(x3, y3)
        results.append(triangle_area2(p1, p2, p3))
    print(" ".join(results))

if __name__ == '__main__':
    main()
